File: dsa-problems/TreesDP/solutions/javascript/TDP-003-subtree-sum-size.js
const readline = require("readline");

class Solution {
  constructor() {
    this.tree = [];
    this.values = [];
    this.subtreeSize = [];
    this.subtreeSum = [];
  }

  computeSubtreeMetrics(n, nodeValues, edges) {
    this.tree = Array.from({ length: n + 1 }, () => []);
    this.values = [0, ...nodeValues];
    this.subtreeSize = Array(n + 1).fill(0);
    this.subtreeSum = Array(n + 1).fill(0);

    for (const [u, v] of edges) {
      this.tree[u].push(v);
      this.tree[v].push(u);
    }

    this.dfs(1, -1);
  }

  dfs(node, parent) {
    this.subtreeSize[node] = 1;
    this.subtreeSum[node] = this.values[node];

    for (const child of this.tree[node]) {
      if (child === parent) continue;

      this.dfs(child, node);
      this.subtreeSize[node] += this.subtreeSize[child];
      this.subtreeSum[node] += this.subtreeSum[child];
    }
  }

  getSubtreeSums() {
    return this.subtreeSum.slice(1);
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const lines = [];
rl.on("line", (line) => {
  lines.push(line);
});

rl.on("close", () => {
  const n = parseInt(lines[0]);
  const values = lines[1].split(" ").map(Number);

  const edges = [];
  for (let i = 2; i < n + 1; i++) {
    const [u, v] = lines[i].split(" ").map(Number);
    edges.push([u, v]);
  }

  const solution = new Solution();
  solution.computeSubtreeMetrics(n, values, edges);

  const sums = solution.getSubtreeSums();
  sums.forEach((s) => console.log(s));
});

File: dsa-problems/TreesDP/solutions/python/TDP-003-subtree-sum-size.py
from typing import List
import sys
sys.setrecursionlimit(300000)

class Solution:
    def __init__(self):
        self.tree = []
        self.values = []
        self.subtree_size = []
        self.subtree_sum = []

    def compute_subtree_metrics(self, n: int, node_values: List[int], edges: List[List[int]]) -> None:
        self.tree = [[] for _ in range(n + 1)]
        self.values = [0] + node_values
        self.subtree_size = [0] * (n + 1)
        self.subtree_sum = [0] * (n + 1)

        for u, v in edges:
            self.tree[u].append(v)
            self.tree[v].append(u)

        self.dfs(1, -1)

    def dfs(self, node: int, parent: int) -> None:
        self.subtree_size[node] = 1
        self.subtree_sum[node] = self.values[node]

        for child in self.tree[node]:
            if child == parent:
                continue

            self.dfs(child, node)
            self.subtree_size[node] += self.subtree_size[child]
            self.subtree_sum[node] += self.subtree_sum[child]

    def get_subtree_sums(self) -> List[int]:
        return self.subtree_sum[1:]

def main():
    lines = sys.stdin.read().strip().split('\n')
    n = int(lines[0])
    values = list(map(int, lines[1].split()))

    edges = []
    for i in range(2, n + 1):
        u, v = map(int, lines[i].split())
        edges.append([u, v])

    solution = Solution()
    solution.compute_subtree_metrics(n, values, edges)

    sums = solution.get_subtree_sums()
    for s in sums:
        print(s)

if __name__ == "__main__":
    main()
